scripts/aws_replicate.py

```diff
@@ -0,0 +1,376 @@
+        mutexLock.acquire()
+        for _, files in file_grps.iteritems():
+            chunk_size = self.global_config.get("chunk_size", 1)
+            idx = 0
+            while idx < len(files):
+                workQueue.put(files[idx : idx + chunk_size])
+                idx = idx + chunk_size
+        mutexLock.release()
+
+        for th in self.thread_list:
+            th.start()
+
+    def run(self,):
+        global TOTAL_PROCESSED_FILES
+
+        start = timeit.default_timer()
+        # Wait for queue to empty
+        while not workQueue.empty():
+            time.sleep(5)
+            current = timeit.default_timer()
+            # running_time = current - start
+            # if int(running_time) % 10== 0:
+            logger.info("=======================================================")
+            logger.info("Total files processed {}".format(TOTAL_PROCESSED_FILES))
+            logger.info("Times in second {}".format(current - start))
+
+        # Notify threads it's time to exit
+        global EXIT_FLAG
+        EXIT_FLAG = 1
+
+        # Wait for all threads to complete
+        for t in self.thread_list:
+            t.join()
+
+        logger.info("Done")
+
+
+def exec_aws_copy(
+    threadName, files, source_bucket, target_bucket, global_config, copied_keys
+):
+    """
+    Call AWS SLI to copy a chunk of  files from a bucket to another bucket.
+    Intergrity check: After each chunk copy, check the returned md5 hashes
+                        with the ones provided in manifest.
+    If not match, re-copy. Log all the success and failure cases
+
+    Args:
+        threadName(str): thread name 
+        files(list): a list of files which should be copied to the same bucket
+        source_bucket(str): source bucket
+        target_bucket(str): target bucket
+        global_config(dict): user config
+        copied_keys(set): set of objects already in target buckets
+    
+    Returns:
+        None
+    """
+
+    config_chunk_size = global_config.get("chunk_size", 1)
+
+    index = 0
+
+    while index < len(files):
+        base_cmd = 'aws s3 cp s3://{} s3://{} --recursive --exclude "*"'.format(
+            source_bucket, target_bucket
+        )
+
+        chunk_size = min(config_chunk_size, len(files) - index)
+
+        execstr = base_cmd
+        for fi in files[index : index + chunk_size]:
+            object_name = "{}/{}".format(fi.get("id"), fi.get("filename"))
+
+            # only copy ones not exist in target bucket
+            if object_name not in copied_keys:
+                execstr += ' --include "{}"'.format(object_name)
+
+        if execstr != base_cmd:
+            subprocess.Popen(shlex.split(execstr + " --quiet")).wait()
+
+        index = index + chunk_size
+
+    global TOTAL_PROCESSED_FILES
+    mutexLock.acquire()
+    TOTAL_PROCESSED_FILES += len(files)
+    mutexLock.release()
+
+
+def process_data(threadName, source_bucket, global_config, copied_objects, q):
```

Comment:
I think this should just be inlined inside the thread class, and then it will also be fine to get the references to a mutex etc from the caller